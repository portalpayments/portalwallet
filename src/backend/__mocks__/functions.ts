const actualFunctions = jest.requireActual("../functions");

// Mock this
export const log = jest.fn();

// Use actual implementation of everything else
export const toUnique = actualFunctions.toUnique;
export const adjustYear = actualFunctions.adjustYear;
export const encodeToBase64 = actualFunctions.encodeToBase64;
export const decodeFromBase64 = actualFunctions.decodeFromBase64;
export const asyncFilter = actualFunctions.asyncFilter;
export const minorUnitsToDecimal = actualFunctions.minorUnitsToDecimal;
export const asyncMap = actualFunctions.asyncMap;
export const sleep = actualFunctions.sleep;
export const stringify = actualFunctions.stringify;
export const isEmpty = actualFunctions.isEmpty;
export const nonNullable = actualFunctions.nonNullable;
export const makePromise = actualFunctions.makePromise;
export const hexToUtf8 = actualFunctions.hexToUtf8;
export const instructionDataToNote = actualFunctions.instructionDataToNote;
