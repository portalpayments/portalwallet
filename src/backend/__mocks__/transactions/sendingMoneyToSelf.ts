import { PublicKey } from "@solana/web3.js";
import {
  USDC_MAINNET_MINT_ACCOUNT,
  MIKES_WALLET,
  SPL_TOKEN_PROGRAM,
  MIKES_USDC_ACCOUNT,
} from "../../constants";

// Mike sending money to his own account
// Which he did once. Mainly this is just to make sure we don't choke on this silly transaction.
export const sendingMoneyToSelf = {
  blockTime: 1662985447,
  meta: {
    err: null,
    fee: 5000,
    innerInstructions: [],
    logMessages: [
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA invoke [1]",
      "Program log: Instruction: TransferChecked",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA consumed 5929 of 200000 compute units",
      "Program TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA success",
    ],
    postBalances: [4620800, 2039280, 146540071068, 934087680],
    postTokenBalances: [
      {
        accountIndex: 1,
        mint: USDC_MAINNET_MINT_ACCOUNT,
        owner: MIKES_WALLET,
        programId: SPL_TOKEN_PROGRAM,
        uiTokenAmount: {
          amount: "25263485",
          decimals: 6,
          uiAmount: 25.263485,
          uiAmountString: "25.263485",
        },
      },
    ],
    preBalances: [4625800, 2039280, 146540071068, 934087680],
    preTokenBalances: [
      {
        accountIndex: 1,
        mint: USDC_MAINNET_MINT_ACCOUNT,
        owner: MIKES_WALLET,
        programId: SPL_TOKEN_PROGRAM,
        uiTokenAmount: {
          amount: "25263485",
          decimals: 6,
          uiAmount: 25.263485,
          uiAmountString: "25.263485",
        },
      },
    ],
    rewards: [],
    status: {
      Ok: null,
    },
  },
  slot: 150330435,
  transaction: {
    message: {
      accountKeys: [
        {
          pubkey: new PublicKey(MIKES_WALLET),
          signer: true,
          source: "transaction",
          writable: true,
        },
        {
          pubkey: new PublicKey(MIKES_USDC_ACCOUNT),
          signer: false,
          source: "transaction",
          writable: true,
        },
        {
          pubkey: new PublicKey(USDC_MAINNET_MINT_ACCOUNT),
          signer: false,
          source: "transaction",
          writable: false,
        },
        {
          pubkey: new PublicKey(SPL_TOKEN_PROGRAM),
          signer: false,
          source: "transaction",
          writable: false,
        },
      ],
      addressTableLookups: null,
      instructions: [
        {
          parsed: {
            info: {
              authority: MIKES_WALLET,
              destination: MIKES_USDC_ACCOUNT,
              mint: USDC_MAINNET_MINT_ACCOUNT,
              source: MIKES_USDC_ACCOUNT,
              tokenAmount: {
                amount: "1000000",
                decimals: 6,
                uiAmount: 1,
                uiAmountString: "1",
              },
            },
            type: "transferChecked",
          },
          program: "spl-token",
          programId: new PublicKey(SPL_TOKEN_PROGRAM),
        },
      ],
      recentBlockhash: "7hATKZN4oKGzmU1x9iMTDJ12Nv6PGfUSoKJkLBtY8Edy",
    },
    signatures: [
      "V7jqDfAeaLniy2crKHHu18LXCivjTczDr1ciFLKNTd7aBA2BC11TrKyzrV8cf9jpgn1UiznuhkuHUJvbAifJQJ6",
    ],
  },
};
