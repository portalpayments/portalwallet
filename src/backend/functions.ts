// Generic JavaScript functions for browser and node
const decoder = new TextDecoder("utf-8");

import base58 from "bs58";

export const toUnique = function <T>(array: Array<T>): Array<T> {
  return Array.from(new Set(array));
};

// Moves dates years back or forward, handling leap years (unlike '+ 1 * YEAR')
export const adjustYear = (date: Date, adjustYears: number): Date => {
  const newDate = new Date(date);
  newDate.setFullYear(date.getFullYear() + adjustYears);
  return newDate;
};

export const encodeToBase64 = (stringToEncode: string): string => {
  return Buffer.from(stringToEncode).toString("base64");
};

export const decodeFromBase64 = (stringToDecode: string): string => {
  return Buffer.from(stringToDecode, "base64").toString("ascii");
};

export const minorUnitsToDecimal = (minorUnits: number, decimal: number) => {
  // Quick note: we must use division instead of multiplicaton here to fix floating point issues:
  // > 420 / 1e6
  // 0.00042
  // > 420 * 1e-6
  // 0.00041999999999999996
  return minorUnits / Math.pow(10, decimal);
};

export const debug = (_ignored) => {};

export const solanaBlocktimeToJSTime = (blockTime: number) => {
  return blockTime * 1000;
};

export const isIncludedCaseInsensitive = (
  string: string,
  substring: string
) => {
  return string.toLowerCase().includes(substring.toLowerCase());
};

// May be unnecessary but 'absolute value' isn't necessaryilt a well known concept
export const removeSign = (number: number) => {
  return Math.abs(number);
};

export const invertNumber = (number: number) => {
  return -number;
};

export const flipZeroAndOne = (number: 0 | 1) => {
  return Number(!number);
};

export const isPositive = (number: number) => {
  return Math.sign(number) === 1;
};

export const dateToISODate = (date: number) => {
  return new Date(date).toISOString().slice(0, 10);
};

export const isoDateToFriendlyName = (isoDate: string) => {
  return new Date(isoDate).toLocaleDateString("en-US", {
    weekday: "long",
    month: "long",
    day: "numeric",
  });
};

export const toLocalTime = (date: number | Date) => {
  if (typeof date === "number") {
    date = new Date(date);
  }
  return new Intl.DateTimeFormat("en-US", {
    hour: "numeric",
    minute: "numeric",
  }).format(date);
};

export const byDateNewestToOldest = (a, b) => {
  if (a.date === b.date) {
    return 0;
  }
  if (a.date < b.date) {
    return 1;
  }
  return -1;
};

export const byDateOldestToNewest = (a, b) => {
  if (a.date === b.date) {
    return 0;
  }
  if (a.date < b.date) {
    return -1;
  }
  return 0;
};

// Adapted from https://github.com/mgenware/node-filter-async/blob/main/src/main.ts
// (which seems to have issues loading its ESM module)
export const asyncFilter = async <T>(
  array: Array<T>,
  filter: (value: T, index: number) => Promise<boolean>
): Promise<Array<T>> => {
  const results: boolean[] = await Promise.all(
    array.map((value, index) => filter(value, index))
  );
  return array.filter((_, index) => results[index]);
};

export const asyncMap = async <ArrayItemType, IteratorReturnType>(
  array: Array<ArrayItemType>,
  iterator: (
    value: ArrayItemType,
    index?: number
  ) => Promise<IteratorReturnType>
): Promise<Array<IteratorReturnType>> => {
  return Promise.all(array.map(iterator));
};

export const sleep = async (timeInMs: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, timeInMs));
};

// 'any' is OK - JSON.stringify itself uses 'any'
export const stringify = (object: any): string => {
  return JSON.stringify(object, null, 2);
};

export const isEmpty = (object: any): boolean => {
  if (object === undefined || object === null) {
    return true;
  }
  return Object.keys(object).length === 0;
};

// https://stackoverflow.com/questions/47632622/typescript-and-filter-boolean
export const nonNullable = <T>(value: T): value is NonNullable<T> => {
  return value !== null && value !== undefined;
};

// Used with mock Jest functions (though you may wish to use mockResolvedValue() instead)
// https://bit.ly/3bB7p4c
export const makePromise = <T>(item: unknown): Promise<T> => {
  return Promise.resolve(item as T);
};

export const log = console.log.bind(console);

// From https://stackoverflow.com/questions/60504945/javascript-encode-decode-utf8-to-hex-and-hex-to-utf8
export const hexToUtf8 = function (string) {
  return decodeURIComponent(
    string
      .replace(/\s+/g, "") // remove spaces
      .replace(/[0-9a-f]{2}/g, "%$&") // add '%' before each 2 characters
  );
};

export const instructionDataToNote = (string: string) => {
  const binaryArray = base58.decode(string);
  return decoder.decode(Buffer.from(binaryArray));
};
