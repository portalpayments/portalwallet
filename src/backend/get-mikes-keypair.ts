// This file is part of Portal Wallet.
//
// Portal Wallet is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, specifically version 2 of the License.
//
// Portal Wallet is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with Portal Wallet. If not, see <https://www.gnu.org/licenses/>.
//
import type { Keypair } from "@solana/web3.js";
import dotenv from "dotenv";
import { getKeypairFromString } from "./wallet";

dotenv.config();

const mikesSecretKey = process.env.MIKES_SECRET_KEY;

if (!mikesSecretKey) {
  throw new Error(`Please set MIKES_SECRET_KEY in .env file`);
}

export const mikesKeypair = getKeypairFromString(mikesSecretKey);
