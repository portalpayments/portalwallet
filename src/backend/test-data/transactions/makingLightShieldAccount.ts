import { PublicKey } from "@solana/web3.js";
import { LIGHT_PROGRAM } from "../../constants";

export const makingLightShieldAccount = {
  blockTime: 1673461335,
  meta: {
    err: null,
    fee: 5000,
    innerInstructions: [],
    logMessages: [
      "Program 11111111111111111111111111111111 invoke [1]",
      "Program 11111111111111111111111111111111 success",
      "Program 2c54pLrGpQdGxJWUAoME6CReBrtDbsx5Tqx4nLZZo6av invoke [1]",
      "Program 2c54pLrGpQdGxJWUAoME6CReBrtDbsx5Tqx4nLZZo6av consumed 3011 of 400000 compute units",
      "Program 2c54pLrGpQdGxJWUAoME6CReBrtDbsx5Tqx4nLZZo6av success",
    ],
    postBalances: [146922253904, 1572960, 1, 1141440, 1009200],
    postTokenBalances: [],
    preBalances: [146923831864, 0, 1, 1141440, 1009200],
    preTokenBalances: [],
    rewards: [],
    status: {
      Ok: null,
    },
  },
  slot: 171973996,
  transaction: {
    message: {
      accountKeys: [
        {
          pubkey: new PublicKey("5FHwkrdxntdK24hgQU8qgBjn35Y1zwhz1GZwCkP2UJnM"),
          signer: true,
          source: "transaction",
          writable: true,
        },
        {
          pubkey: new PublicKey("7ePbj5NeQsjQQ8pRd6H62XfktbytpCicbEAzhiyNYjkL"),
          signer: false,
          source: "transaction",
          writable: true,
        },
        {
          pubkey: new PublicKey("11111111111111111111111111111111"),
          signer: false,
          source: "transaction",
          writable: false,
        },
        {
          pubkey: new PublicKey(LIGHT_PROGRAM),
          signer: false,
          source: "transaction",
          writable: false,
        },
        {
          pubkey: new PublicKey("SysvarRent111111111111111111111111111111111"),
          signer: false,
          source: "transaction",
          writable: false,
        },
      ],
      addressTableLookups: null,
      instructions: [
        {
          parsed: {
            info: {
              base: "5FHwkrdxntdK24hgQU8qgBjn35Y1zwhz1GZwCkP2UJnM",
              lamports: 1572960,
              newAccount: "7ePbj5NeQsjQQ8pRd6H62XfktbytpCicbEAzhiyNYjkL",
              owner: LIGHT_PROGRAM,
              seed: "lightLookup",
              source: "5FHwkrdxntdK24hgQU8qgBjn35Y1zwhz1GZwCkP2UJnM",
              space: 98,
            },
            type: "createAccountWithSeed",
          },
          program: "system",
          programId: new PublicKey("11111111111111111111111111111111"),
        },
        {
          accounts: [
            "5FHwkrdxntdK24hgQU8qgBjn35Y1zwhz1GZwCkP2UJnM",
            "7ePbj5NeQsjQQ8pRd6H62XfktbytpCicbEAzhiyNYjkL",
            "SysvarRent111111111111111111111111111111111",
          ],
          data: "11111111g14VBrqegAkHgkiR4WUVZojNLdG7TTaQN3UbQ3dLWakrsiKABgiq715d2wfx4ZLkoGNTmrJ9ShLaeZmJV9psFNfkNs",
          programId: new PublicKey(LIGHT_PROGRAM),
        },
      ],
      recentBlockhash: "B7rHvsrW7NBmYAi8bHDTFGgXXYDtXB3vaG5V3ynjR4ad",
    },
    signatures: [
      "5egDQUg2pvuTyvdyEnUVmQg3zXr2ZrLWAC33rS7rDxmWJGuprohY8Bj25PAfpwyN3iT6KU5eFg9AojBgC1dRxsq8",
    ],
  },
  version: "legacy",
};
