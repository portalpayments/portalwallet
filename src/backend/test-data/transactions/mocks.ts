export const MOCK_SENDER_PUBLIC_KEY =
  "AfpdXGBdFiyQfXdf4x3yYWiJgPqp6xBNvy3UwHyuLgGn";
export const MOCK_RECIPIENT_PUBLIC_KEY =
  "2DCC44zP2BDxTKzgi4yYScEJmbyhW2G8aYCQUrEcABts";
