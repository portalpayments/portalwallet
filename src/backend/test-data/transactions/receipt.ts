import { MIKES_WALLET } from "../../constants";

export const rawReceipt = {
  props: {
    pageProps: {
      order: {
        restaurantId: "mtJcsdB9ETLJdmjrOLff",
        discountItem: {},
        payingPubKey: MIKES_WALLET,
        squarePaymentIds: ["LNkSvZUuVADHFaAXHpFoS3uqEt9YY"],
        discountBreakdown: {
          "solana-or-not-financial-advice-tee": {
            appliedDiscountPubKey: "",
            appliedDiscountNFTName: "",
            appliedDiscountUSD: 0,
            isMaxDiscount: false,
          },
        },
        squarePaymentId: "LNkSvZUuVADHFaAXHpFoS3uqEt9YY",
        refundStatus: "notRefunded",
        tip: 0,
        discount: 0,
        subtotal: 32,
        paymentStatus: "paid",
        creationTime: 1667678976911,
        items: [
          {
            taxIds: ["YX7PHIKISGZU53WNHSFE6BT6"],
            productId: "solana-or-not-financial-advice-tee",
            toAmount: 0,
            candyMachineId: "",
            orderQuantity: 1,
            specialInstructions: "",
            externalId: "VJ6CQI2ABZUOKZMVO44KYDB2",
            image:
              "https://items-images-production.s3.us-west-2.amazonaws.com/files/48b95aff5cbfc2087daeac70e1037728f4180867/original.png",
            type: "product",
            productName: "Solana | Not Financial Advice Tee",
            unitPrice: 32,
            addOns: [
              {
                stock: 87,
                name: "Medium, White",
                slug: "solana-or-not-financial-advice-tee-medium-white",
                id: "solana-or-not-financial-advice-tee-medium-white",
                categorySlug: "solana-or-not-financial-advice-tee-variants",
                externalId: "4AGMD3ZOHMU3J5V5XGCGUMFO",
                sku: "695202K",
                price: {
                  value: 32,
                  currencyCode: "usd",
                },
                status: "publish",
              },
            ],
          },
        ],
        cashier: "Tessa",
        squareOrderId: "hhe5I60wV1ZqHrU4ge5vFrR03QSZY",
        orderReference: "A64u27MpYAYNJvpgnKjEHwTuJBBNfNphco4f46S752ZA",
        tax: 0,
        functionError: null,
        requireSquareSync: true,
        txOutcome: {
          err: "",
          outcome: true,
          tx: "2Wgzyv1fFFiF4jd8ckPvSJa2eRBHF7pj3wbeTpEzMvqzpfx1jpUgWUZbJW9h915rxWccNqZ9ksFjP7PVVckArZtX",
        },
        total: 32,
        nativeTotal: 0,
        name: "Admin",
        status: "completed",
        attemptedPaymentNum: 1,
        email: "",
        table: "",
        phoneNumber: "",
        fiatTotal: 32,
        gateway: "solana_counter",
        id: "qqS5qxxEjMg7mSup0rBI",
      },
      shop: {
        notifications: [
          {
            priority: 1,
            time: 1667844487802,
            text: "Wallet balance is low for Breakpoint Prod Fee Payer",
            id: "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
          },
          {
            text: "Wallet balance is low for Breakpoint Prod Fee Payer",
            priority: 1,
            id: "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
            time: 1667934392665,
          },
        ],
        settings: {
          solanaAddress: "4iDRFnp2N4UAsZEePHAxs7ozBanQcGtLYd12HG2HJm4s",
          requireSquareSync: true,
          require2FA: false,
          bitcoinAddress: "",
          activeOrdering: false,
          acceptCard: false,
          requireTable: false,
          acceptSolanaInAppWalletPayment: false,
          acceptSolanaCounterWalletPayment: false,
          acceptSolana: true,
          currency: {
            id: "usd",
            name: "USD",
          },
          exchangeFeePercentageCard: 0,
          exchangeWallet: {
            walletPublicKey: "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
            walletLabel: "Breakpoint Prod Fee Payer",
          },
          isExchangeActive: true,
          tippingPercentage: 0,
          acceptBitcoin: false,
          transactionFeesPaidFor: true,
          exchangeFeePercentageCash: 0,
          solanaWalletFeePayer: {
            walletPublicKey: "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
            walletLabel: "Breakpoint Prod Fee Payer",
          },
          taxPercentage: 0,
          taxes: [],
          acceptCash: true,
        },
        contact: {
          number: "",
          name: "Solana Spaces Breakpoint",
        },
        slug: "spacesbreakpoint",
        logoUri: {
          firebase:
            "https://firebasestorage.googleapis.com/v0/b/decaf-c57a3.appspot.com/o/images%2Fusers%2FMZcVyOrfU8dWUd7U5lKVWDAJX1k1%2Flogo%2F4b8ce112-0a84-4f26-a3f7-6b094f46635a-Solana_logo.png?alt=media",
          cdn: "https://ik.imagekit.io/r3y94jb5zai/firebasePROD/o/images%2Fusers%2FMZcVyOrfU8dWUd7U5lKVWDAJX1k1%2Flogo%2F4b8ce112-0a84-4f26-a3f7-6b094f46635a-Solana_logo.png?alt=media",
        },
        name: "Solana Spaces Breakpoint",
      },
      dev: false,
    },
    __N_SSG: true,
  },
  page: "/receipt/[receiptId]",
  query: {
    receiptId: "XgVU1qK4i4zXKanjHZpr",
  },
  buildId: "s_U2jOX4q6ShosslyfVMJ",
  isFallback: false,
  gsp: true,
  scriptLoader: [],
};
