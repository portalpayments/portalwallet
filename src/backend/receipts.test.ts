import {
  getDecafReceiptMessage,
  getRawReceiptFromMessage,
  getReceiptForSimpleTransaction,
  getReceiptSummaryFromRawReceipt,
} from "./receipts";
import { stringify, log } from "./functions";
import { mikesKeypair } from "./get-mikes-keypair";
import { ThreadMemberScope } from "@dialectlabs/sdk";
import { rawReceipt } from "./test-data/transactions/receipt";
import {
  DECAF_APP,
  getMintAddressBySymbol,
  MIKES_WALLET,
  SECONDS,
} from "./constants";

jest.mock("./functions");
jest.mock("./html-extract");

describe(`dialect`, () => {
  test(
    `getting messages`,
    async () => {
      // Date Mike bought a t shirt
      const transactionDate = new Date("November 05, 2022 20:10:02").valueOf();

      const receiptMessage = await getDecafReceiptMessage(
        mikesKeypair,
        transactionDate
      );

      expect(receiptMessage).toEqual({
        author: {
          address: DECAF_APP,
          scopes: ["ADMIN", "WRITE"],
        },
        timestamp: expect.anything(),
        text: "Thank you for your order! You can find your receipt here: https://www.decaf.so/receipt/XgVU1qK4i4zXKanjHZpr",
      });
    },
    10 * SECONDS
  );

  test(`Getting receipts from messages`, async () => {
    const rawReceipt = await getRawReceiptFromMessage({
      author: {
        address: DECAF_APP,
        scopes: [ThreadMemberScope.ADMIN, ThreadMemberScope.WRITE],
      },
      timestamp: expect.anything(),
      text: "Thank you for your order! You can find your receipt here: https://www.decaf.so/receipt/XgVU1qK4i4zXKanjHZpr",
    });

    expect(rawReceipt).toMatchSnapshot({
      buildId: expect.any(String),
    });
  });

  test(`Getting receiptSummary from raw receipts`, async () => {
    const receiptSummary = getReceiptSummaryFromRawReceipt(rawReceipt);
    expect(receiptSummary).toMatchObject({
      items: [
        {
          name: "Not Financial Advice Tee",
          price: 32,
          quantity: 1,
        },
      ],
      shop: "Solana Spaces Breakpoint",
    });
  });

  test(
    `Adds receipts to transaction that have them`,
    async () => {
      const simpleTransaction = {
        id: "2Wgzyv1fFFiF4jd8ckPvSJa2eRBHF7pj3wbeTpEzMvqzpfx1jpUgWUZbJW9h915rxWccNqZ9ksFjP7PVVckArZtX",
        date: 1667679002000,
        status: true,
        networkFee: 10000,
        direction: 0,
        amount: 32000000,
        currency: getMintAddressBySymbol("USDC"),
        from: MIKES_WALLET,
        to: "4iDRFnp2N4UAsZEePHAxs7ozBanQcGtLYd12HG2HJm4s",
        memo: "qqS5qxxEjMg7mSup0rBI",
      };

      const receipt = await getReceiptForSimpleTransaction(
        mikesKeypair,
        simpleTransaction.memo,
        simpleTransaction.date
      );
      expect(receipt).toMatchObject({});
    },
    // TODO, really slow but may be a Dialect network issue
    15 * SECONDS
  );
});
