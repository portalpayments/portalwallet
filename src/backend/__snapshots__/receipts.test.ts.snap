// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dialect Getting receipts from messages 1`] = `
{
  "buildId": Any<String>,
  "gsp": true,
  "isFallback": false,
  "page": "/receipt/[receiptId]",
  "props": {
    "__N_SSG": true,
    "pageProps": {
      "dev": false,
      "order": {
        "attemptedPaymentNum": 1,
        "cashier": "Tessa",
        "creationTime": 1667678976911,
        "discount": 0,
        "discountBreakdown": {
          "solana-or-not-financial-advice-tee": {
            "appliedDiscountNFTName": "",
            "appliedDiscountPubKey": "",
            "appliedDiscountUSD": 0,
            "isMaxDiscount": false,
          },
        },
        "discountItem": {},
        "email": "",
        "fiatTotal": 32,
        "functionError": null,
        "gateway": "solana_counter",
        "id": "qqS5qxxEjMg7mSup0rBI",
        "items": [
          {
            "addOns": [
              {
                "categorySlug": "solana-or-not-financial-advice-tee-variants",
                "externalId": "4AGMD3ZOHMU3J5V5XGCGUMFO",
                "id": "solana-or-not-financial-advice-tee-medium-white",
                "name": "Medium, White",
                "price": {
                  "currencyCode": "usd",
                  "value": 32,
                },
                "sku": "695202K",
                "slug": "solana-or-not-financial-advice-tee-medium-white",
                "status": "publish",
                "stock": 87,
              },
            ],
            "candyMachineId": "",
            "externalId": "VJ6CQI2ABZUOKZMVO44KYDB2",
            "image": "https://items-images-production.s3.us-west-2.amazonaws.com/files/48b95aff5cbfc2087daeac70e1037728f4180867/original.png",
            "orderQuantity": 1,
            "productId": "solana-or-not-financial-advice-tee",
            "productName": "Solana | Not Financial Advice Tee",
            "specialInstructions": "",
            "taxIds": [
              "YX7PHIKISGZU53WNHSFE6BT6",
            ],
            "toAmount": 0,
            "type": "product",
            "unitPrice": 32,
          },
        ],
        "name": "Admin",
        "nativeTotal": 0,
        "orderReference": "A64u27MpYAYNJvpgnKjEHwTuJBBNfNphco4f46S752ZA",
        "payingPubKey": "5FHwkrdxntdK24hgQU8qgBjn35Y1zwhz1GZwCkP2UJnM",
        "paymentStatus": "paid",
        "phoneNumber": "",
        "refundStatus": "notRefunded",
        "requireSquareSync": true,
        "restaurantId": "mtJcsdB9ETLJdmjrOLff",
        "squareOrderId": "hhe5I60wV1ZqHrU4ge5vFrR03QSZY",
        "squarePaymentId": "LNkSvZUuVADHFaAXHpFoS3uqEt9YY",
        "squarePaymentIds": [
          "LNkSvZUuVADHFaAXHpFoS3uqEt9YY",
        ],
        "status": "completed",
        "subtotal": 32,
        "table": "",
        "tax": 0,
        "tip": 0,
        "total": 32,
        "txOutcome": {
          "err": "",
          "outcome": true,
          "tx": "2Wgzyv1fFFiF4jd8ckPvSJa2eRBHF7pj3wbeTpEzMvqzpfx1jpUgWUZbJW9h915rxWccNqZ9ksFjP7PVVckArZtX",
        },
      },
      "shop": {
        "contact": {
          "name": "Solana Spaces Breakpoint",
          "number": "",
        },
        "logoUri": {
          "cdn": "https://ik.imagekit.io/r3y94jb5zai/firebasePROD/o/images%2Fusers%2FMZcVyOrfU8dWUd7U5lKVWDAJX1k1%2Flogo%2F4b8ce112-0a84-4f26-a3f7-6b094f46635a-Solana_logo.png?alt=media",
          "firebase": "https://firebasestorage.googleapis.com/v0/b/decaf-c57a3.appspot.com/o/images%2Fusers%2FMZcVyOrfU8dWUd7U5lKVWDAJX1k1%2Flogo%2F4b8ce112-0a84-4f26-a3f7-6b094f46635a-Solana_logo.png?alt=media",
        },
        "name": "Solana Spaces Breakpoint",
        "notifications": [
          {
            "id": "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
            "priority": 1,
            "text": "Wallet balance is low for Breakpoint Prod Fee Payer",
            "time": 1667844487802,
          },
          {
            "id": "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
            "priority": 1,
            "text": "Wallet balance is low for Breakpoint Prod Fee Payer",
            "time": 1667934392665,
          },
        ],
        "settings": {
          "acceptBitcoin": false,
          "acceptCard": false,
          "acceptCash": true,
          "acceptSolana": true,
          "acceptSolanaCounterWalletPayment": false,
          "acceptSolanaInAppWalletPayment": false,
          "activeOrdering": false,
          "bitcoinAddress": "",
          "currency": {
            "id": "usd",
            "name": "USD",
          },
          "exchangeFeePercentageCard": 0,
          "exchangeFeePercentageCash": 0,
          "exchangeWallet": {
            "walletLabel": "Breakpoint Prod Fee Payer",
            "walletPublicKey": "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
          },
          "isExchangeActive": true,
          "require2FA": false,
          "requireSquareSync": true,
          "requireTable": false,
          "solanaAddress": "4iDRFnp2N4UAsZEePHAxs7ozBanQcGtLYd12HG2HJm4s",
          "solanaWalletFeePayer": {
            "walletLabel": "Breakpoint Prod Fee Payer",
            "walletPublicKey": "DhF84xpim3cPL53vcNRNfJByBKNn73jTaAghSpe3DwRH",
          },
          "taxPercentage": 0,
          "taxes": [],
          "tippingPercentage": 0,
          "transactionFeesPaidFor": true,
        },
        "slug": "spacesbreakpoint",
      },
    },
  },
  "query": {
    "receiptId": "XgVU1qK4i4zXKanjHZpr",
  },
  "scriptLoader": [],
}
`;
